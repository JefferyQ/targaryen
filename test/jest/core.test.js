/**
 * Jest test definition to test targaryen Jest integration.
 */

'use strict';

const targaryen = require('../../plugins/jest');

expect.extend({
  toBeAllowed: targaryen.toBeAllowed
});

test('getDebugDatabase()', () => {
  const emptyRules = {rules: {}};
  const database = targaryen.getDebugDatabase(emptyRules, {});

  expect(database.debug).toBe(true);
});

test('getDatabase()', () => {
  const emptyRules = {rules: {}};
  const database = targaryen.getDatabase(emptyRules, {});

  expect(database.debug).toBe(false);
});

describe('generic matchers', () => {
  test('toBeAllowed', () => {
    const rules = {
      rules: {
        user: {
          $uid: {
            '.read': 'auth.uid !== null',
            '.write': 'auth.uid === $uid'
          }
        }
      }
    };
    const initialData = {};

    // NOTE: Create a database with debug set to true for detailed errors
    const database = targaryen.getDebugDatabase(rules, initialData);

    expect(() => {
      expect(database.as(null).read('/user')).not.toBeAllowed();
    }).not.toThrow();

    expect(() => {
      expect(database.as(null).read('/user')).toBeAllowed();
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      expect(database.as({uid: '1234'}).write('/user/1234', {
        name: 'Anna'
      })).toBeAllowed();
    }).not.toThrow();

    expect(() => {
      expect(database.as({uid: '1234'}).write('/user/1234', {
        name: 'Anna'
      })).not.toBeAllowed();
    }).toThrowErrorMatchingSnapshot();
  });
});

