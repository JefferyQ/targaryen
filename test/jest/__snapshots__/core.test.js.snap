// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generic matchers toBeAllowed 1`] = `
"Expected operation to be [32mallowed[39m but it was [31mdenied[39m

Attempt to read /user as null.

/user: read <no rules>

No .read rule allowed the operation.
read was denied."
`;

exports[`generic matchers toBeAllowed 2`] = `
"Expected operation to be [32mdenied[39m but it was [31mallowed[39m

Attempt to write /user/1234 as {\\"uid\\":\\"1234\\"}.
New Value: \\"{
  \\"name\\": \\"Anna\\"
}\\".

/user/1234: write \\"auth.uid === $uid\\"  => true
  auth.uid === $uid  [=> true]
  using [
    $uid = \\"1234\\"
    auth = {\\"uid\\":\\"1234\\"}
    auth.uid = \\"1234\\"
  ]

write was allowed."
`;
